// node-kakao/dist_esm/packet/struct/wrap/user.js
import { UserType } from '../../../user/user-type.js';

function structToChannelUserInfo(struct) {
    return {
        userId: struct.userId,
        userType: struct.type,
        nickname: struct.nickName,
        countryIso: struct.countryIso,
        statusMessage: struct.statusMessage,
        suspended: struct.suspended,
        linkedServices: struct.linkedServices,
        profileURL: struct.profileImageUrl,
        fullProfileURL: struct.fullProfileImageUrl,
        originalProfileURL: struct.originalProfileImageUrl,
        ut: struct.ut,
        accountId: struct.accountId,
    };
}
function structToOpenChannelUserInfo(struct) {
    return {
        userId: struct.userId,
        linkId: struct.pli,
        openToken: struct.opt,
        perm: struct.mt,
        userType: struct.type,
        nickname: struct.nickName,
        profileURL: struct.pi,
        fullProfileURL: struct.fpi,
        originalProfileURL: struct.opi,
    };
}
function structToOpenLinkChannelUserInfo(struct) {
    return {
        userId: struct.userId,
        linkId: struct.pli,
        openToken: struct.opt,
        perm: struct.lmt,
        userType: UserType.OPEN_PROFILE,
        nickname: struct.nn,
        profileURL: struct.pi,
        fullProfileURL: struct.fpi,
        originalProfileURL: struct.opi,
        privilege: struct.pv,
        profileType: struct.ptp,
    };
}
function structToOpenLinkKickedUserInfo(struct) {
    return {
        nickname: struct.nickName,
        userId: struct.userId,
        profileURL: struct.pi,
        kickedChannelId: struct.c,
    };
}

export { structToChannelUserInfo, structToOpenChannelUserInfo, structToOpenLinkChannelUserInfo, structToOpenLinkKickedUserInfo };
