// node-kakao/dist_esm/api/axios-web-client.js
import Axios from 'axios';
import { convertToFormData } from './web-api-util.js';
import FormData from 'form-data';

class AxiosWebClient {
    constructor(scheme, host, _decorator) {
        this.scheme = scheme;
        this.host = host;
        this._decorator = _decorator;
    }
    get url() {
        return `${this.scheme}://${this.host}`;
    }
    toApiURL(path) {
        return `${this.url}/${path}`;
    }
    fillHeader(header) {
        var _a;
        header['Host'] = this.host;
        (_a = this._decorator) === null || _a === void 0 ? void 0 : _a.fillHeader(header);
    }
    buildAxiosReqData(method, header) {
        const headers = {};
        this.fillHeader(headers);
        const reqData = {
            headers,
            method,
            transformResponse: (data) => data,
            responseType: 'arraybuffer',
            maxContentLength: 100000000,
            maxBodyLength: 100000000
        };
        if (header)
            Object.assign(headers, header);
        return reqData;
    }
    async request(method, path, form, headers) {
        const reqData = this.buildAxiosReqData(method, headers);
        reqData.url = this.toApiURL(path);
        if (form) {
            const formData = convertToFormData(form);
            reqData.data = formData.toString();
        }
        const res = await Axios.request(reqData);
        if (res.status !== 200) {
            throw new Error(`Web request failed with status ${res.status} ${res.statusText}`);
        }
        return res.data;
    }
    async requestMultipart(method, path, form, headers) {
        const reqData = this.buildAxiosReqData(method, headers);
        reqData.url = this.toApiURL(path);
        if (form) {
            const formData = this.convertToMultipart(form);
            Object.assign(reqData.headers, formData.getHeaders());
            reqData.data = formData.getBuffer();
        }
        const res = await Axios.request(reqData);
        if (res.status !== 200) {
            throw new Error(`Web request failed with status ${res.status} ${res.statusText}`);
        }
        return res.data;
    }
    convertToMultipart(form) {
        const formData = new FormData();
        for (const [key, value] of Object.entries(form)) {
            if (value && value.value && value.options) {
                const file = value;
                formData.append(key, Buffer.from(file.value), file.options);
            }
            else {
                formData.append(key, value + '');
            }
        }
        return formData;
    }
}

export { AxiosWebClient };
