// node-kakao/dist/talk/talk-channel-list.js
/*
 * Created on Wed Jan 27 2021
 *
 * Copyright (c) storycraft. Licensed under the MIT Licence.
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "../event", "../util", "./openlink", "./channel"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.TalkChannelList = void 0;
    const event_1 = require("../event");
    const util_1 = require("../util");
    const openlink_1 = require("./openlink");
    const channel_1 = require("./channel");
    /**
     * Manage normal channels and open channels
     */
    class TalkChannelList extends event_1.TypedEmitter {
        /**
         * Construct managed channel list
         * @param {TalkSession} session
         * @param {ClientDataLoader} loader
         * @param {TalkNormalChannel[]} normalList
         * @param {TalkOpenChannel[]} openList
         * @param {InformedOpenLink[]} clientLinkList
         */
        constructor(session, loader, normalList = [], openList = [], clientLinkList = []) {
            super();
            this._normal = new channel_1.TalkNormalChannelList(session, loader, normalList);
            this._open = new openlink_1.TalkOpenChannelList(session, loader, openList, clientLinkList);
        }
        get size() {
            return this._normal.size + this._open.size;
        }
        /**
         * Normal channel list
         */
        get normal() {
            return this._normal;
        }
        /**
         * Open channel list
         */
        get open() {
            return this._open;
        }
        get(channelId) {
            return this._normal.get(channelId) || this._open.get(channelId);
        }
        all() {
            const normalIter = this._normal.all();
            const openIter = this._open.all();
            return new util_1.ChainedIterator(normalIter, openIter);
        }
        pushReceived(method, data, parentCtx) {
            const ctx = new event_1.EventContext(this, parentCtx);
            if (method === 'MSG') {
                const msgData = data;
                if (!this.get(msgData.chatId)) {
                    let list;
                    if ('li' in msgData) {
                        list = this._open;
                    }
                    else {
                        list = this._normal;
                    }
                    list.addChannel({ channelId: msgData.chatId }).then((res) => {
                        if (!res.success)
                            return;
                        ctx.emit('channel_added', res.result);
                        this._normal.pushReceived(method, data, ctx);
                        this._open.pushReceived(method, data, ctx);
                    });
                    return;
                }
            }
            this._normal.pushReceived(method, data, ctx);
            this._open.pushReceived(method, data, ctx);
        }
        /**
         * Initialize TalkChannelList using channelList.
         * @param {TalkChannelList} talkChannelList
         * @param {LoginData<NormalChannelData | OpenChannelData>[]} channelList
         */
        static async initialize(talkChannelList, channelList = []) {
            const [normalList, openList] = TalkChannelList.mapChannelList(channelList);
            await Promise.all([
                channel_1.TalkNormalChannelList.initialize(talkChannelList._normal, normalList),
                openlink_1.TalkOpenChannelList.initialize(talkChannelList._open, openList),
            ]);
            return talkChannelList;
        }
        /**
         * Split normal channel and open channel
         *
         * @param {LoginData<NormalChannelData | OpenChannelData>[]} channelList
         * @return {[LoginData<NormalChannelData>[], LoginData<OpenChannelData>[]]}
         */
        static mapChannelList(channelList) {
            const normalList = [];
            const openList = [];
            channelList.forEach((data) => {
                if ('linkId' in data.channel) {
                    openList.push(data);
                }
                else {
                    normalList.push(data);
                }
            });
            return [normalList, openList];
        }
    }
    exports.TalkChannelList = TalkChannelList;
});
