// node-kakao/dist_esm/openlink/index.js
export { OpenChannelInfo } from './open-channel-info.js';
export { OpenLinkProfile } from './open-link-profile.js';
export { OpenChannelType, OpenChannelUserPerm, OpenLinkType, OpenProfileType } from './open-link-type.js';

var KnownLinkPrivilegeMask;
(function (KnownLinkPrivilegeMask) {
    KnownLinkPrivilegeMask[KnownLinkPrivilegeMask["URL_SHARABLE"] = 2] = "URL_SHARABLE";
    KnownLinkPrivilegeMask[KnownLinkPrivilegeMask["REPORTABLE"] = 4] = "REPORTABLE";
    KnownLinkPrivilegeMask[KnownLinkPrivilegeMask["PROFILE_EDITABLE"] = 8] = "PROFILE_EDITABLE";
    KnownLinkPrivilegeMask[KnownLinkPrivilegeMask["ANY_PROFILE_ALLOWED"] = 32] = "ANY_PROFILE_ALLOWED";
    KnownLinkPrivilegeMask[KnownLinkPrivilegeMask["USE_PASS_CODE"] = 64] = "USE_PASS_CODE";
    KnownLinkPrivilegeMask[KnownLinkPrivilegeMask["BLINDABLE"] = 128] = "BLINDABLE";
    KnownLinkPrivilegeMask[KnownLinkPrivilegeMask["NON_SPECIAL_LINK"] = 512] = "NON_SPECIAL_LINK";
    KnownLinkPrivilegeMask[KnownLinkPrivilegeMask["USE_BOT"] = 1024] = "USE_BOT";
})(KnownLinkPrivilegeMask || (KnownLinkPrivilegeMask = {}));

export { KnownLinkPrivilegeMask };
