// node-kakao/dist_esm/openlink/open-link-type.js
var OpenLinkType;
(function (OpenLinkType) {
    OpenLinkType[OpenLinkType["PROFILE"] = 1] = "PROFILE";
    OpenLinkType[OpenLinkType["CHANNEL"] = 2] = "CHANNEL";
})(OpenLinkType || (OpenLinkType = {}));
var OpenChannelUserPerm;
(function (OpenChannelUserPerm) {
    OpenChannelUserPerm[OpenChannelUserPerm["OWNER"] = 1] = "OWNER";
    OpenChannelUserPerm[OpenChannelUserPerm["NONE"] = 2] = "NONE";
    OpenChannelUserPerm[OpenChannelUserPerm["MANAGER"] = 4] = "MANAGER";
    OpenChannelUserPerm[OpenChannelUserPerm["BOT"] = 8] = "BOT";
})(OpenChannelUserPerm || (OpenChannelUserPerm = {}));
var OpenProfileType;
(function (OpenProfileType) {
    OpenProfileType[OpenProfileType["MAIN"] = 1] = "MAIN";
    OpenProfileType[OpenProfileType["KAKAO_ANON"] = 2] = "KAKAO_ANON";
    OpenProfileType[OpenProfileType["KAKAO_ANON_2"] = 4] = "KAKAO_ANON_2";
    OpenProfileType[OpenProfileType["UNKNOWN_1"] = 8] = "UNKNOWN_1";
    OpenProfileType[OpenProfileType["OPEN_PROFILE"] = 16] = "OPEN_PROFILE";
})(OpenProfileType || (OpenProfileType = {}));
var OpenChannelType;
(function (OpenChannelType) {
    OpenChannelType[OpenChannelType["UNKNOWN"] = 0] = "UNKNOWN";
    OpenChannelType[OpenChannelType["DIRECT"] = 1] = "DIRECT";
    OpenChannelType[OpenChannelType["GROUP"] = 2] = "GROUP";
})(OpenChannelType || (OpenChannelType = {}));

export { OpenChannelType, OpenChannelUserPerm, OpenLinkType, OpenProfileType };
