// node-kakao/dist_esm/talk/block/index.js
import { KnownDataStatusCode } from '../../request/index.js';

class TalkBlockSession {
    constructor(_session) {
        this._session = _session;
    }
    async blockUser(user, type = TalkBlockType.BLOCK) {
        const res = await this._session.request('BLADDITEM', {
            'l': [user.userId],
            'ts': [type],
        });
        return { success: res.status === KnownDataStatusCode.SUCCESS, status: res.status };
    }
    async blockPlusUser(plusUser, type = TalkBlockType.BLOCK) {
        const res = await this._session.request('BLADDITEM', {
            'pl': [plusUser.userId],
            'pts': [type],
        });
        return { success: res.status === KnownDataStatusCode.SUCCESS, status: res.status };
    }
    async unblockUser(user) {
        const res = await this._session.request('BLDELITEM', {
            'l': [user.userId],
        });
        return { success: res.status === KnownDataStatusCode.SUCCESS, status: res.status };
    }
    async unblockPlusUser(plusUser) {
        const res = await this._session.request('BLDELITEM', {
            'pl': [plusUser.userId],
        });
        return { success: res.status === KnownDataStatusCode.SUCCESS, status: res.status };
    }
}
var TalkBlockType;
(function (TalkBlockType) {
    TalkBlockType[TalkBlockType["BLOCK"] = 0] = "BLOCK";
    TalkBlockType[TalkBlockType["BLOCK_HIDE_PROFILE"] = 1] = "BLOCK_HIDE_PROFILE";
})(TalkBlockType || (TalkBlockType = {}));

export { TalkBlockSession, TalkBlockType };
